### DISCUSSION TITLE
Article Filter: How do I add `msg.isImportant` to whitelist keywords and Labels to blacklist keywords?

### DISCUSSION THREAD

**Discussion by Crehet** [martinrotter/rssguard#988] [Q&A] [answered]
I've tried a few ways, but as always, I dunno how.
By the way, I not trying to only fetch specific Articles and `Ignoring` the rest. But rather, simply add `Important` and `Labels` to every already fetched Articles and the ones yet to be fetched.

```js
var whitelist = [
  'Pepa', 'Max-Steel'
];

var blacklist = [
  'John Wick 1', 'John Wick 2'
];

function filterMessage() {
  if (whitelist.some(i => msg.title.indexOf(i) != -1)) {    //is this fine?
   msg.isImportant = true;
   return MessageObject.Accept;
 } else {
   if (blacklist.some(i => msg.title.indexOf(i) != -1)) {    //is this fine?
    msg.assignLabel(msg.findLabelId('Boring/Futa/Furry/3D/CGI'));    //are slash `/` even allowed here?
   }
   return MessageObject.Accept;
  }
}
```

**Comment by Mitecon** [comment, marked as answer]
Firstly, why not have two labels? Something like:
```
Accepted
Unwanted
```
This way, you can see what's matched your wanted list and then everything else that matches your unwanted list is still there for review.

So it looks like you want this to be marked Important:
```
var whitelist = [
  'Pepa', 'Max-Steel'
];
```
and this to be in a label:
```
var blacklist = [
  'John Wick 1', 'John Wick 2'
];
```
Then what do you want to do with everything else? Do you _only_ want the articles matching either of the conditions above in the database? Or do you want the above to happen and _also_ keep all other articles? This part is unclear to me.


Anyway, the main problem part is the 'else if' statement. The syntax is slightly incorrect but enough to stop it from working altogether.

You need to change this:
```
 } else {
   if (blacklist.some(i => msg.title.indexOf(i) != -1)) {
```
to this:
```
 } else if
   (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))) {
```
I changed the condition to be case-insensitive but you can adjust if necessary.

So the full working filter would be:
```
var whitelist = [
  'Pepa',
  'Max-Steel'
];

var blacklist = [
  'John Wick 1',
  'John Wick 2'
];

function filterMessage() {
  if (whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) {
   msg.isImportant = true;
   msg.assignLabel(msg.findLabelId('Accepted'));
   return MessageObject.Accept;
 } else if
   (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))) {
    msg.assignLabel(msg.findLabelId('Unwanted'));
    return MessageObject.Accept;
   } else {
   return MessageObject.Ignore;
  }
}
```
At the very end is:
```
   return MessageObject.Ignore;
```
but if you still want to keep everything else, obviously change it to:
```
   return MessageObject.Accept;
```
and those articles will be neither marked Important, nor in a label but they'll still be there in the feed.

I'd also avoid using slashes in label names. Usually you'd need to escape special characters with a backslash  but why risk it? Javascript requires escaping special characters but if you can create a label with special characters then it could become a bit messy.

In the example above I call the label 'Unwanted'. Probably best to keep it simple.

You can also obviously remove the line:
```
   msg.assignLabel(msg.findLabelId('Accepted'));
```
if you really don't want that.

**Reply by Crehet** [reply]
> Firstly, why not have two labels?

I'd like to, but its difficult and tacky to see two or more than three labels separated by a `,`. I'd prefer to just stick to one for now until things get better on that end.

>Then what do you want to do with everything else? Do you only want the articles matching either of the conditions above in the database? Or do you want the above to happen and also keep all other articles? This part is unclear to me.

I really appreciate how well you understood my words. Yes, I wanted to match either of the conditions while also keeping all other articles. So, no rejecting nor deleting/removing but accepting and keeping and finally appending labels and `Important` to articles.

> You need to change this:
```js
 } else {
   if (blacklist.some(i => msg.title.indexOf(i) != -1)) {
```

> to this:
```js
 } else if
   (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))) {
```

This one:
```js
 } else if
   (!(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1))) {
    msg.assignLabel(msg.findLabelId('Unwanted'));
```

Was not working as I expected it to be. It was appending the label to everything else but blacklisted/whitelisted keywords, so I slightly changed it to see if it worked this way:

```js
(blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) {
```
And worked as I desired. I dunno why `(i` at the beginning was not making it work for me.

So, the final result ended with the above as the only difference.

```js
var whitelist = [
  'Pepa',
  'Max-Steel'
];

var blacklist = [
  'John Wick 1',
  'John Wick 2'
];

function filterMessage() {
  if (whitelist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) {
   msg.isImportant = true;
//   msg.assignLabel(msg.findLabelId('Accepted'));    // don't need this line
   return MessageObject.Accept;
 } else if
   (blacklist.some(i => msg.title.toLowerCase().indexOf(i.toLowerCase()) != -1)) {    // removed '(i' to make it work.
    msg.assignLabel(msg.findLabelId('Un/wan/ted'));    // wasn't sure if labels with slashes like: `Un/wan/ted` would work, but it did.
    return MessageObject.Accept;
//   } else {
//   return MessageObject.Ignore;    // don't need the above line and this one.
  }
}

**Comment by Mitecon** [comment]
> I'd like to, but its difficult and tacky to see two or more than three labels separated by a `,`. I'd prefer to just stick to one for now until things get better on that end.

Wait - what are you calling a 'label'? Labels aren't separated by a comma. Are you talking about variables?:
```
var whitelist = [
  'Pepa',
  'Max-Steel'
];
```

> And worked as I desired. I dunno why `(i` at the beginning was not making it work for me.

Yes, this highlights the importance of doing your own testing instead of blindly using something someone else told you would work - just like I did. Oops!

Glad you figured out how to adapt it to your own situation.

The '(!' at the beginning of the line is to negate, meaning 'NOT'. So I obviously got it a bit wrong there while I was copy/pasting one of my own filters.

**Reply by Crehet** [reply]
>Wait - what are you calling a 'label'? Labels aren't separated by a comma. Are you talking about variables?

No. To see what I was referring to; press <kbd>H</kbd> to see `Headers` > right click over the Articles Header > Checkmark `Assigned labels`. Now add a _few_ labels to any Article just for testing, you'll see they will be shown separated by a `,` i.e: _Done,Unwated,Accepted_

 >The '(!' at the beginning of the line is to negate, meaning 'NOT'. So I obviously got it a bit wrong there while I was copy/pasting one of my own filters.

I'm interested on how you've got your filters... Might be a decent idea to submit a PR for adding more examples to the [Article Filtering section.](https://github.com/martinrotter/rssguard/blob/master/resources/docs/Documentation.md#fltr)

**Comment by Mitecon** [comment]
I see what you're referring to now. That's just a visual thing which I suppose doesn't bother me. Also, I'm using a DevBuild version which has the label names disabled for now so all I see in those columns are numbers.

It doesn't affect anything other than whether or not you like seeing it that way but you figured out a solution you like best for that anyway.

I got my filters by simply learning just like you. I was a bit stuck when I first started but filters are powerful and once you get the idea of how to construct them it becomes easier. All you really need is to learn the basics of Javascript and you don't really need to go much beyond that. In fact, what I do is take the time to create a filter, do lots of testing, then pretty much copy and paste it for other filters and adapt as needed. Filters become simpler when you get the hang of them. However, mine look extremely complex now but it's just because they contain many of the different elements.

It's basically:
```
Set variables (as many as you want)
Define conditions (these will only be either TRUE or FALSE)
Define actions (either Accept if TRUE or not if FALSE)
```
The first filter is the hardest - then it's just copy/paste.

It also helps to use something that can help you visualise things. There are many online Javascript editors. Now I use Kate (part of KDE Plasma), set the highlighting to Javascript and then it will colourise everything and will highlight matching parentheses and curly braces. You'll find your own workflow that suits you.

I'm by no means an expert - I'm still learning myself.

**Comment by martinrotter** [comment]
Hi guys, so is the discussion solved? If not please feel free to ask me questions about each single issue you have.
